napari/_qt/containers/qt_layer_model.py

```diff
@@ -0,0 +1,70 @@
+from qtpy.QtCore import QModelIndex, QSize, Qt
+from qtpy.QtGui import QImage
+
+from ...layers import Layer
+from .qt_list_model import QtListModel
+
+
+class QtLayerListModel(QtListModel[Layer]):
+    LayerRole = Qt.UserRole
+    ThumbnailRole = Qt.UserRole + 1
+
+    def data(self, index: QModelIndex, role: Qt.ItemDataRole):
+        """Return data stored under ``role`` for the item at ``index``."""
+        layer = self.getItem(index)
+        if role == Qt.DisplayRole:  # used for item text
+            return layer.name
+        if role == Qt.TextAlignmentRole:  # alignment of the text
+            return Qt.AlignCenter
+        if role == Qt.EditRole:  # used to populate line edit when editing
+            return layer.name
+        if role == Qt.ToolTipRole:  # for tooltip
+            return layer.name
+        if role == Qt.CheckStateRole:  # the "checked" state of this item
+            return Qt.Checked if layer.visible else Qt.Unchecked
```

Comment:
I've experimented with this branch, and it works, thanks @tlambert03!

I needed to work around `QtLayerListModel` connected to a combo box though: model's `CheckStateRole` for a particular row depends on layer visibility for that row, which clashes with combo box check state (all items appear to be checked). I overrided `data()` to return non-null values on "relevant" roles, not sure if this is the right approach.

Reply:
>  I overrided data() to return non-null values on "relevant" roles, not sure if this is the right approach.

i'd probably need to see a bit more code (not sure I'm following what you mean about the combo box here) ... but overriding `data()` to customize stuff for your needs is absolutely a good/standard approach.  So no worries there!

Reply:
> i'd probably need to see a bit more code (not sure I'm following what you mean about the combo box here)

```python
box = QComboBox()
model = napari_viewer.window.qt_viewer.layers.model()
box.setModel(model)
```

![image](https://user-images.githubusercontent.com/1649961/114359613-723c5400-9b74-11eb-9f93-ec62006799c8.png)

I find it strange because [QComboBox doc](https://doc.qt.io/qt-5/qcombobox.html) says:

> For the text and icon in the combobox label, the data in the model that has the Qt::DisplayRole and Qt::DecorationRole is used.